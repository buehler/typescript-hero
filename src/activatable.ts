export default interface Activatable {
  setup(): void;
  start(): void;
  stop(): void;
  dispose(): void;
}
