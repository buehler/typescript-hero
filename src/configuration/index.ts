import { inject, injectable } from 'inversify';
import { Observable, Subject } from 'rxjs';
import { ExtensionContext, workspace } from 'vscode';

import iocSymbols from '../ioc-symbols';
import DocumentOutlineConfig from './document-outline-config';

const sectionKey = 'typescriptHero';

@injectable()
export default class Configuration {
  public readonly codeOutline: DocumentOutlineConfig = new DocumentOutlineConfig();

  private readonly _configurationChanged: Subject<void> = new Subject();

  public get configurationChanged(): Observable<void> {
    return this._configurationChanged;
  }

  constructor(
    @inject(iocSymbols.extensionContext) context: ExtensionContext,
  ) {
    context.subscriptions.push(
      workspace.onDidChangeConfiguration(() => this._configurationChanged.next()),
    );
  }

  public parseableLanguages(): string[] {
    return [
      'typescript',
      'typescriptreact',
      'javascript',
      'javascriptreact',
    ];
  }

  public verbosity(): 'error' | 'warn' | 'info' | 'debug' {
    const verbosity = workspace.getConfiguration(sectionKey).get<'error' | 'warn' | 'info' | 'debug'>('verbosity', 'warn');
    if (['error', 'warn', 'info', 'debug'].indexOf(verbosity) < 0) {
      return 'warn';
    }
    return verbosity;
  }
}
