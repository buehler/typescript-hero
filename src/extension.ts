import 'reflect-metadata';
import {TypeScriptHero} from './TypeScriptHero';
import * as vscode from 'vscode';
import {Injector} from './IoC';

let extension: vscode.Disposable;

export function activate(context: vscode.ExtensionContext) {
    Injector.bind<vscode.ExtensionContext>('context').toConstantValue(context);
    extension = Injector.get(TypeScriptHero);
}

export function deactivate() {
    extension.dispose();
    extension = null;
}
