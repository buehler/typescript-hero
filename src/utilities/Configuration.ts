import * as vscode from 'vscode';

const sectionKey = 'typescriptHero';

export class Configuration {
    public static get pathStringDelimiter(): string {
        return vscode.workspace.getConfiguration(sectionKey).get<string>('pathStringDelimiter');
    }

    public static get includeNodeModules(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('includeNodeModules');
    }

    public static get includeTypings(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('includeTypings');
    }

    public static get refreshOnSave(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('refreshOnSave');
    }

    // public static get organizeOnSave(): boolean {
    //     return vscode.workspace.getConfiguration(sectionKey).get<boolean>('organizeOnSave');
    // }

    public static get debug(): boolean {
        return vscode.workspace.getConfiguration(sectionKey).get<boolean>('debug');
    }
}
