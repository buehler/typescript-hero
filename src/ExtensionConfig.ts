import { ImportLocation, TsImportOptions } from './models/TsImportOptions';
import { LogLevel } from './utilities/Logger';
import { injectable } from 'inversify';
import { workspace } from 'vscode';

const sectionKey = 'typescriptHero';

@injectable()
export class ExtensionConfig {
    private resolverConfig: ResolverConfig = new ResolverConfig();
    private restartDebuggerConfig: RestartDebuggerConfig = new RestartDebuggerConfig();

    public get logLevel(): LogLevel {
        let optString = workspace.getConfiguration(sectionKey).get<string>('verbosity');
        switch (optString) {
            case 'Nothing':
                return LogLevel.Nothing;
            case 'Errors':
                return LogLevel.Errors;
            case 'All':
                return LogLevel.All;
            default:
                return LogLevel.Warnings;
        }
    }

    public get resolver(): ResolverConfig {
        return this.resolverConfig;
    }

    public get restartDebugger(): RestartDebuggerConfig {
        return this.restartDebuggerConfig;
    }
}

class ResolverConfig {
    public get insertSpaceBeforeAndAfterImportBraces(): boolean {
        return workspace.getConfiguration(sectionKey).get<boolean>('resolver.insertSpaceBeforeAndAfterImportBraces');
    }

    public get pathStringDelimiter(): string {
        return workspace.getConfiguration(sectionKey).get<string>('resolver.pathStringDelimiter');
    }

    public get ignorePatterns(): string[] {
        return workspace.getConfiguration(sectionKey).get<string[]>('resolver.ignorePatterns');
    }

    public get multiLineWrapThreshold(): number {
        return workspace.getConfiguration(sectionKey).get<number>('resolver.multiLineWrapThreshold');
    }

    public get newImportLocation(): ImportLocation {
        let configString = workspace.getConfiguration(sectionKey).get<string>('resolver.newImportLocation');
        return ImportLocation[configString];
    }

    public get tabSize(): number {
        return workspace.getConfiguration().get<number>('editor.tabSize');
    }

    public get importOptions(): TsImportOptions {
        return {
            multiLineWrapThreshold: this.multiLineWrapThreshold,
            pathDelimiter: this.pathStringDelimiter,
            spaceBraces: this.insertSpaceBeforeAndAfterImportBraces,
            tabSize: this.tabSize
        };
    }
}

class RestartDebuggerConfig {
    public get watchFolders(): string[] {
        return workspace.getConfiguration(sectionKey).get<string[]>('restartDebugger.watchFolders');
    }

    public get active(): boolean {
        return workspace.getConfiguration(sectionKey).get<boolean>('restartDebugger.active');
    }
}
