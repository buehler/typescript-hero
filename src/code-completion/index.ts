import { inject, injectable } from 'inversify';
import { DeclarationInfo, TypescriptParser } from 'typescript-parser';
import {
  CancellationToken,
  commands,
  CompletionItem,
  CompletionItemProvider,
  ExtensionContext,
  languages,
  Position,
  TextDocument,
  workspace,
} from 'vscode';

import Activatable from '../activatable';
import Configuration from '../configuration';
import DeclarationManager from '../declarations/declaration-manager';
import iocSymbols, { ImportManagerProvider } from '../ioc-symbols';
import { Logger } from '../utilities/logger';
import { getDeclarationsFilteredByImports, getItemKind, getScriptKind } from '../utilities/utility-functions';

const REGEX_COMMENT = /^\s*(?:\/\/|\/\*|\*\/|\*)/;

/**
 * Extension that provides code completion for typescript and javascript files.
 * Uses the calculated index to provide information.
 *
 * @export
 * @class CodeCompletion
 * @implements {Activatable}
 * @implements {CompletionItemProvider}
 */
@injectable()
export class CodeCompletion implements Activatable, CompletionItemProvider {

  constructor(
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.configuration) private config: Configuration,
    @inject(iocSymbols.importManager) private importManagerProvider: ImportManagerProvider,
    @inject(iocSymbols.declarationManager) private declarationManager: DeclarationManager,
    @inject(iocSymbols.parser) private parser: TypescriptParser,
  ) { }

  public setup(): void {
    this.logger.debug('[CodeCompletion] Setting up CodeCompletion.');
    this.logger.debug(
      '[CodeCompletion] Registering for languages.',
      { languages: this.config.parseableLanguages() },
    );

    for (const lang of this.config.parseableLanguages()) {
      this.context.subscriptions.push(languages.registerCompletionItemProvider(lang, this));
    }

    this.context.subscriptions.push(
      commands.registerCommand(
        'typescriptHero.codeCompletion.executeIntellisenseItem',
        (document: TextDocument, declaration: DeclarationInfo) =>
          this.executeIntellisenseItem(document, declaration),
      ),
    );
  }

  public start(): void {
    this.logger.info('[CodeCompletion] Starting up CodeCompletion.');
  }

  public stop(): void {
    this.logger.info('[CodeCompletion] Stopping CodeCompletion.');
  }

  public dispose(): void {
    this.logger.debug('[CodeCompletion] Disposing CodeCompletion.');
  }

  public async provideCompletionItems(
    document: TextDocument,
    position: Position,
    token: CancellationToken,
  ): Promise<CompletionItem[] | null> {
    const index = this.declarationManager.getIndexForFile(document.uri);
    const actualWorkspace = workspace.getWorkspaceFolder(document.uri);

    if (!index || !actualWorkspace || !index.indexReady) {
      this.logger.info('[CodeCompletion] index not ready or no workspace folder selected');
      return null;
    }

    const wordAtPosition = document.getWordRangeAtPosition(position);
    const lineText = document.lineAt(position.line).text;

    let searchWord = '';

    if (wordAtPosition && wordAtPosition.start.character < position.character) {
      const word = document.getText(wordAtPosition);
      searchWord = word.substr(0, position.character - wordAtPosition.start.character);
    }

    if (
      !searchWord ||
      token.isCancellationRequested ||
      (lineText.substring(0, position.character).match(/["'`]/g) || []).length % 2 === 1 ||
      REGEX_COMMENT.test(lineText) ||
      lineText.startsWith('import ') ||
      new RegExp(`(?:\w*\.)+${searchWord}`).test(lineText.substring(0, position.character))
    ) {
      this.logger.debug(
        '[CodeCompletion] did not match criteria to provide intellisense',
        { searchWord, lineText, indexReady: index.indexReady },
      );
      return null;
    }

    this.logger.debug(`[CodeCompletion] provide code completion for "${searchWord}"`);
    const profiler = this.logger.startTimer();

    const parsed = await this.parser.parseSource(document.getText(), getScriptKind(document.fileName));
    const declarations = getDeclarationsFilteredByImports(
      index.declarationInfos,
      document.fileName,
      parsed.imports,
      actualWorkspace.uri.fsPath,
    )
      .filter(o => !parsed.declarations.some(d => d.name === o.declaration.name))
      .filter(o => !parsed.usages.some(d => d === o.declaration.name));

    const items: CompletionItem[] = [];
    for (const declaration of declarations.filter(
      o => o.declaration.name.toLowerCase().indexOf(searchWord.toLowerCase()) >= 0)
    ) {
      const item = new CompletionItem(declaration.declaration.name, getItemKind(declaration.declaration));

      item.detail = declaration.from;
      item.command = {
        arguments: [document, declaration],
        title: 'Execute intellisense insert',
        command: 'typescriptHero.codeCompletion.executeIntellisenseItem',
      };
      if (this.config.codeCompletion.completionSortMode(document.uri) === 'bottom') {
        item.sortText = `9999-${declaration.declaration.name}`;
      }
      items.push(item);
    }

    profiler.done({ message: `[CodeCompletion] calculated code completions` });
    return items;
  }

  /**
   * Executes a intellisense item that provided a document and a declaration to add.
   * Does make the calculation of the text edits async.
   *
   * @private
   * @param {TextDocument} document
   * @param {DeclarationInfo} declaration
   * @returns {Promise<void>}
   * @memberof CodeCompletion
   */
  private async executeIntellisenseItem(document: TextDocument, declaration: DeclarationInfo): Promise<void> {
    this.logger.info(
      '[CodeCompletion] Execute code completion action.',
      { specifier: declaration.declaration.name, library: declaration.from },
    );
    await (await this.importManagerProvider(document)).addDeclarationImport(declaration).commit();
  }
}
