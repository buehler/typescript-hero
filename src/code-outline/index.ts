import { inject, injectable } from 'inversify';
import { Subscription } from 'rxjs';
import { File, TypescriptParser } from 'typescript-parser';
import {
  Disposable,
  Event,
  EventEmitter,
  ExtensionContext,
  ProviderResult,
  TreeDataProvider,
  window,
  workspace,
} from 'vscode';

import Activatable from '../activatable';
import Configuration from '../configuration';
import iocSymbols from '../ioc-symbols';
import { Logger } from '../utilities/Logger';
import { getScriptKind } from '../utilities/utilityFunctions';
import BaseStructureTreeItem from './base-structure-tree-item';
import DeclarationStructureTreeItem from './declaration-structure-tree-item';
import DisabledStructureTreeItem from './disabled-structure-tree-item';
import { ImportsStructureTreeItem } from './imports-structure-tree-item';
import NotParseableStructureTreeItem from './not-parseable-structure-tree-item';
import ResourceStructureTreeItem from './resource-structure-tree-item';

@injectable()
export default class CodeOutline implements Activatable, TreeDataProvider<BaseStructureTreeItem> {
  private _onDidChangeTreeData: EventEmitter<BaseStructureTreeItem | undefined>;

  private subscription: Subscription;
  private disposables: Disposable[] = [];
  private documentCache?: File;

  public get onDidChangeTreeData(): Event<BaseStructureTreeItem | undefined> {
    return this._onDidChangeTreeData.event;
  }

  constructor(
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.configuration) private config: Configuration,
    @inject(iocSymbols.parser) private parser: TypescriptParser,
  ) { }

  public setup(): void {
    this.logger.debug('Setting up DocumentOutline.');
    this.subscription = this.config.configurationChanged.subscribe(() => {
      if (this.config.codeOutline.isEnabled() && !this.disposables) {
        this.start();
      } else if (!this.config.codeOutline.isEnabled() && this.disposables) {
        this.stop();
      }
    });
  }

  public start(): void {
    if (!this.config.codeOutline.isEnabled()) {
      this.logger.debug(`Not starting DocumentOutline. It's disabled by config.`);
      return;
    }
    this.logger.debug('Starting up DocumentOutline.');
    this._onDidChangeTreeData = new EventEmitter<BaseStructureTreeItem | undefined>();
    this.disposables.push(window.registerTreeDataProvider('codeOutline', this));
    this.context.subscriptions.push(this._onDidChangeTreeData);
    this.context.subscriptions.push(window.onDidChangeActiveTextEditor(() => this.activeWindowChanged()));
    this.context.subscriptions.push(workspace.onDidSaveTextDocument(() => this.activeWindowChanged()));
  }

  public stop(): void {
    if (this.config.codeOutline.isEnabled()) {
      this.logger.debug(`Not stopping DocumentOutline. It's enabled by config.`);
      return;
    }
    this.logger.debug('Stopping DocumentOutline.');
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }

  public dispose(): void {
    this.logger.debug('Disposing DocumentOutline.');
    if (this.subscription) {
      this.subscription.unsubscribe();
      delete this.subscription;
    }
    for (const disposable of this.disposables) {
      disposable.dispose();
    }
    this.disposables = [];
  }

  public getTreeItem(element: BaseStructureTreeItem): BaseStructureTreeItem {
    return element;
  }

  public async getChildren(element?: BaseStructureTreeItem): Promise<ProviderResult<BaseStructureTreeItem[]>> {
    if (!window.activeTextEditor) {
      return [];
    }

    if (!this.config.codeOutline.isEnabled()) {
      return [new DisabledStructureTreeItem()];
    }

    if (!this.config.parseableLanguages().some(
      lang => lang === window.activeTextEditor!.document.languageId,
    )) {
      return [new NotParseableStructureTreeItem()];
    }

    if (!this.documentCache) {
      try {
        this.documentCache = await this.parser.parseSource(
          window.activeTextEditor.document.getText(),
          getScriptKind(window.activeTextEditor.document.fileName),
        );
      } catch (e) {
        this.logger.error(
          `[DocumentOutline] document could not be parsed, error: ${e}`,
        );
        return [];
      }
    }

    if (!element) {
      const items: BaseStructureTreeItem[] = [];
      if (this.documentCache.imports && this.documentCache.imports.length) {
        items.push(new ImportsStructureTreeItem(this.documentCache, this.context));
      }
      items.push(...this.documentCache.resources.map(r => new ResourceStructureTreeItem(r, this.context)));
      items.push(...this.documentCache.declarations.map(d => new DeclarationStructureTreeItem(d, this.context)));
      return items;
    }
    return element.getChildren();
  }

  /**
     * Method that recalculates the current document when the active window changed.
     *
     * @private
     *
     * @memberof DocumentSymbolStructureExtension
     */
  private activeWindowChanged(): void {
    this.logger.debug('[DocumentOutline] activeWindowChanged, reparsing');
    this.documentCache = undefined;
    this._onDidChangeTreeData.fire();
  }
}
