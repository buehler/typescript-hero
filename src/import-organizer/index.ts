import { inject, injectable } from 'inversify';
import { DeclarationInfo, TypescriptParser } from 'typescript-parser';
import { commands, ExtensionContext, window, workspace } from 'vscode';

import Activatable from '../activatable';
import Configuration from '../configuration';
import DeclarationManager from '../declarations/declaration-manager';
import iocSymbols, { ImportManagerProvider } from '../ioc-symbols';
import { Logger } from '../utilities/logger';
import { getDeclarationsFilteredByImports, getScriptKind } from '../utilities/utility-functions';
import ImportQuickPickItem from './import-quick-pick-item';

type DeclarationsForImportOptions = { cursorSymbol: string, documentSource: string, documentPath: string };

@injectable()
export default class ImportOrganizer implements Activatable {
  constructor(
    @inject(iocSymbols.extensionContext) private context: ExtensionContext,
    @inject(iocSymbols.logger) private logger: Logger,
    @inject(iocSymbols.configuration) private config: Configuration,
    @inject(iocSymbols.importManager) private importManagerProvider: ImportManagerProvider,
    @inject(iocSymbols.declarationManager) private declarationManager: DeclarationManager,
    @inject(iocSymbols.parser) private parser: TypescriptParser,
  ) { }

  public setup(): void {
    this.logger.debug('[ImportOrganizer] Setting up ImportOrganizer.');
    this.context.subscriptions.push(
      commands.registerTextEditorCommand('typescriptHero.imports.organize', () => this.organizeImports()),
    );
    this.context.subscriptions.push(
      commands.registerTextEditorCommand('typescriptHero.index.addImport', () => this.addImport()),
    );
    this.context.subscriptions.push(
      commands.registerTextEditorCommand(
        'typescriptHero.index.addImportUnderCursor',
        () => this.addImportUnderCursor(),
      ),
    );
    this.context.subscriptions.push(
      workspace.onWillSaveTextDocument((event) => {
        if (!this.config.imports.organizeOnSave(event.document.uri)) {
          this.logger.debug(
            '[ImportOrganizer] OrganizeOnSave is deactivated through config',
          );
          return;
        }
        if (this.config.parseableLanguages().indexOf(event.document.languageId) < 0) {
          this.logger.debug(
            '[ImportOrganizer] OrganizeOnSave not possible for given language',
            { language: event.document.languageId },
          );
          return;
        }
        if (this.config.imports.excludeFromSave(event.document.uri)
          .some(regexString => new RegExp(regexString).test(event.document.fileName))) {
          this.logger.debug(
            '[ImportOrganizer] OrganizeOnSave is skipped through excludeFromSave for this file',
            { language: event.document.languageId },
          );
          return;
        }

        this.logger.info(
          '[ImportOrganizer] OrganizeOnSave for file',
          { file: event.document.fileName },
        );
        event.waitUntil(
          this.importManagerProvider(event.document).then(manager => manager.organizeImports().calculateTextEdits()),
        );
      }),
    );
  }

  public start(): void {
    this.logger.info('[ImportOrganizer] Starting up ImportOrganizer.');
  }

  public stop(): void {
    this.logger.info('[ImportOrganizer] Stopping ImportOrganizer.');
  }

  public dispose(): void {
    this.logger.debug('[ImportOrganizer] Disposing ImportOrganizer.');
  }

  /**
   * Organizes the imports of the actual document. Sorts and formats them correctly.
   *
   * @private
   * @returns {Promise<boolean>}
   *
   * @memberof ImportResolveExtension
   */
  private async organizeImports(): Promise<void> {
    if (!window.activeTextEditor) {
      return;
    }
    try {
      this.logger.debug(
        '[ImportOrganizer] Organize the imports in the document',
        { file: window.activeTextEditor.document.fileName },
      );
      const ctrl = await this.importManagerProvider(window.activeTextEditor.document);
      await ctrl.organizeImports().commit();
    } catch (e) {
      this.logger.error(
        '[ImportOrganizer] Imports could not be organized, error: %s',
        e,
        { file: window.activeTextEditor.document.fileName },
      );
    }
  }

  /**
   * Add an import from the whole list. Calls the vscode gui, where the user can
   * select a symbol to import.
   *
   * @private
   * @returns {Promise<void>}
   *
   * @memberof ResolveExtension
   */
  private async addImport(): Promise<void> {
    if (!window.activeTextEditor) {
      return;
    }
    const index = this.declarationManager.getIndexForFile(window.activeTextEditor.document.uri);
    if (!index || !index.indexReady) {
      this.showCacheWarning();
      return;
    }
    try {
      const selectedItem = await window.showQuickPick(
        index.declarationInfos.map(o => new ImportQuickPickItem(o)),
        { placeHolder: 'Add import to document:' },
      );
      if (selectedItem) {
        this.logger.info(
          '[ImportOrganizer] Add import to document',
          { specifier: selectedItem.declarationInfo.declaration.name, library: selectedItem.declarationInfo.from },
        );
        this.addImportToDocument(selectedItem.declarationInfo);
      }
    } catch (e) {
      this.logger.error(
        '[ImportOrganizer] Import could not be added to document',
        { file: window.activeTextEditor.document.fileName, error: e.toString() },
      );
      window.showErrorMessage('The import cannot be completed, there was an error during the process.');
    }
  }

  /**
   * Add an import from the whole list. Calls the vscode gui, where the user can
   * select a symbol to import.
   *
   * @private
   * @returns {Promise<void>}
   *
   * @memberof ImportResolveExtension
   */
  private async addImportUnderCursor(): Promise<void> {
    if (!window.activeTextEditor) {
      return;
    }
    const index = this.declarationManager.getIndexForFile(window.activeTextEditor.document.uri);
    if (!index || !index.indexReady) {
      this.showCacheWarning();
      return;
    }
    try {
      const selectedSymbol = this.getSymbolUnderCursor();
      this.logger.debug('[ImportOrganizer] Add import for symbol under cursor', { selectedSymbol });
      if (!!!selectedSymbol) {
        return;
      }
      const resolveItems = await this.getDeclarationsForImport({
        cursorSymbol: selectedSymbol,
        documentSource: window.activeTextEditor.document.getText(),
        documentPath: window.activeTextEditor.document.fileName,
      });

      if (resolveItems.length < 1) {
        this.logger.info(
          '[ImportOrganizer] The symbol was not found or is already imported',
          { selectedSymbol },
        );
        window.showInformationMessage(
          `The symbol '${selectedSymbol}' was not found in the index or is already imported.`,
        );
      } else if (resolveItems.length === 1 && resolveItems[0].declaration.name === selectedSymbol) {
        this.logger.info(
          '[ImportOrganizer] Add import to document',
          {
            specifier: resolveItems[0].declaration.name,
            library: resolveItems[0].from,
          },
        );
        this.addImportToDocument(resolveItems[0]);
      } else {
        const selectedItem = await window.showQuickPick(
          resolveItems.map(o => new ImportQuickPickItem(o)), { placeHolder: 'Multiple declarations found:' },
        );
        if (selectedItem) {
          this.logger.info(
            '[ImportOrganizer] Add import to document',
            {
              specifier: selectedItem.declarationInfo.declaration.name,
              library: selectedItem.declarationInfo.from,
            },
          );
          this.addImportToDocument(selectedItem.declarationInfo);
        }
      }
    } catch (e) {
      this.logger.error(
        '[ImportOrganizer] Import could not be added to document.',
        { file: window.activeTextEditor.document.fileName, error: e.toString() },
      );
      window.showErrorMessage('The import cannot be completed, there was an error during the process.');
    }
  }

  /**
   * Calculates the possible imports for a given document source with a filter for the given symbol.
   * Returns a list of declaration infos that may be used for select picker or something.
   *
   * @private
   * @param {DeclarationsForImportOptions} {cursorSymbol, documentSource, documentPath}
   * @returns {(Promise<DeclarationInfo[] | undefined>)}
   *
   * @memberof ImportResolveExtension
   */
  private async getDeclarationsForImport(
    { cursorSymbol, documentSource, documentPath }: DeclarationsForImportOptions,
  ): Promise<DeclarationInfo[]> {
    if (!window.activeTextEditor) {
      return [];
    }

    const index = this.declarationManager.getIndexForFile(window.activeTextEditor.document.uri);
    const rootFolder = workspace.getWorkspaceFolder(window.activeTextEditor.document.uri);

    if (!index || !index.indexReady || !rootFolder) {
      return [];
    }

    this.logger.debug(
      '[ImportOrganizer] Calculate possible imports for document',
      { cursorSymbol, file: documentPath },
    );

    const parsedSource = await this.parser.parseSource(documentSource, getScriptKind(documentPath));
    const activeDocumentDeclarations = parsedSource.declarations.map(o => o.name);
    const declarations = getDeclarationsFilteredByImports(
      index.declarationInfos,
      documentPath,
      parsedSource.imports,
      rootFolder.uri.fsPath,
    ).filter(o => o.declaration.name.startsWith(cursorSymbol));

    return [
      ...declarations.filter(o => o.from.startsWith('/')),
      ...declarations.filter(o => !o.from.startsWith('/')),
    ].filter(o => activeDocumentDeclarations.indexOf(o.declaration.name) === -1);
  }

  /**
   * Returns the string under the cursor.
   *
   * @private
   * @returns {string}
   *
   * @memberof ImportResolveExtension
   */
  private getSymbolUnderCursor(): string {
    const editor = window.activeTextEditor;
    if (!editor) {
      return '';
    }
    const selection = editor.selection;
    const word = editor.document.getWordRangeAtPosition(selection.active);

    return word && !word.isEmpty ? editor.document.getText(word) : '';
  }

  /**
   * Effectifely adds an import quick pick item to a document
   *
   * @private
   * @param {DeclarationInfo} declaration
   * @returns {Promise<boolean>}
   *
   * @memberof ImportResolveExtension
   */
  private async addImportToDocument(declaration: DeclarationInfo): Promise<boolean> {
    if (!window.activeTextEditor) {
      return false;
    }
    const ctrl = await this.importManagerProvider(window.activeTextEditor.document);
    return ctrl.addDeclarationImport(declaration).commit();
  }

  /**
   * Shows a user warning if the resolve index is not ready yet.
   *
   * @private
   *
   * @memberof ImportResolveExtension
   */
  private showCacheWarning(): void {
    this.logger.warn(
      '[ImportOrganizer] Index was not ready or not index for this file found',
    );
    window.showWarningMessage('Please wait a few seconds longer until the symbol cache has been build.');
  }
}
