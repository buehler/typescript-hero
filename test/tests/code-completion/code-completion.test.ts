import { join } from 'path';
import { stub } from 'sinon';
import { CancellationTokenSource, Position, TextDocument, Uri, window, workspace } from 'vscode';

import { CodeCompletion } from '../../../src/code-completion';
import Configuration from '../../../src/configuration';
import DeclarationManager from '../../../src/declarations/declaration-manager';
import ioc from '../../../src/ioc';
import iocSymbols from '../../../src/ioc-symbols';
import { expect, waitForIndexReady } from '../setup';

describe('CodeCompletion', () => {

  const rootPath = workspace.workspaceFolders![0].uri.fsPath;
  const token = new CancellationTokenSource().token;
  const file = Uri.file(join(rootPath, 'code-completion', 'completions.ts'));
  let document: TextDocument;
  let extension: CodeCompletion;
  let declarations: DeclarationManager;
  let config: Configuration;

  before(async () => {
    document = await workspace.openTextDocument(file);
    await window.showTextDocument(document);

    declarations = ioc.get<DeclarationManager>(iocSymbols.declarationManager);
    config = ioc.get<Configuration>(iocSymbols.configuration);
    extension = new CodeCompletion(
      ioc.get(iocSymbols.extensionContext),
      ioc.get(iocSymbols.logger),
      config,
      ioc.get(iocSymbols.importManager),
      declarations,
      ioc.get(iocSymbols.parser),
    );

    await waitForIndexReady(declarations.getIndexForFile(document.uri)!);
  });

  describe('provideCompletionItems()', () => {

    it('should provide a completion list', async () => {
      const result = await extension.provideCompletionItems(document, new Position(0, 17), token);
      expect(result).to.matchSnapshot();
    });

    it('should provide a completion list for interfaces', async () => {
      const result = await extension.provideCompletionItems(document, new Position(1, 17), token);
      expect(result).to.matchSnapshot();
    });

    it('should provide a completion list for default imports', async () => {
      const result = await extension.provideCompletionItems(document, new Position(2, 6), token);
      expect(result).to.matchSnapshot();
    });

    it('should not provide a completion list for non ready indices', async () => {
      const mock = stub(declarations, 'getIndexForFile').callsFake(() => ({ indexReady: false }));
      try {
        const result = await extension.provideCompletionItems(document, new Position(0, 0), token);
        expect(result).to.be.null;
      } finally {
        mock.restore();
      }
    });

    it('should not provide a completion list for undefined workspaces', async () => {
      const mock = stub(workspace, 'getWorkspaceFolder').callsFake(() => undefined);
      try {
        const result = await extension.provideCompletionItems(document, new Position(0, 0), token);
        expect(result).to.be.null;
      } finally {
        mock.restore();
      }
    });

    it('should not provide a completion list for undefined indices', async () => {
      const mock = stub(declarations, 'getIndexForFile').callsFake(() => undefined);
      try {
        const result = await extension.provideCompletionItems(document, new Position(0, 0), token);
        expect(result).to.be.null;
      } finally {
        mock.restore();
      }
    });

    it('should not provide a completion list for typing in a string', async () => {
      const result = await extension.provideCompletionItems(document, new Position(3, 18), token);
      expect(result).to.be.null;
    });

    it('should not provide a completion list for typing after a "."', async () => {
      const result = await extension.provideCompletionItems(document, new Position(4, 24), token);
      expect(result).to.be.null;
    });

    it('should not provide a completion list for typing in a comment', async () => {
      const result = await extension.provideCompletionItems(document, new Position(5, 20), token);
      expect(result).to.be.null;
    });

    it('should not provide a completion list for typing in an import', async () => {
      const result = await extension.provideCompletionItems(document, new Position(6, 24), token);
      expect(result).to.be.null;
    });

    it('should not provide a completion for the own file', async () => {
      const result = await extension.provideCompletionItems(document, new Position(9, 5), token);
      expect(result).to.matchSnapshot();
    });

    it('should provide a completion list bottom sorted', async () => {
      const mock = stub(config.codeCompletion, 'completionSortMode').callsFake(() => 'bottom');
      try {
        const result = await extension.provideCompletionItems(document, new Position(0, 17), token);
        expect(result).to.matchSnapshot();
      } finally {
        mock.restore();
      }
    });

  });

});
