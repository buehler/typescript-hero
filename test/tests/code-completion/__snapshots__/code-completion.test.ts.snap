// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeCompletion provideCompletionItems() should not provide a completion for the own file 1`] = `Array []`;

exports[`CodeCompletion provideCompletionItems() should provide a completion list 1`] = `
Array [
  Object {
    "detail": "/code-completion/classes",
    "documentation": undefined,
    "filterText": undefined,
    "insertText": undefined,
    "kind": "Class",
    "label": "CodeCompletionClass1",
    "sortText": undefined,
    "textEdit": undefined,
  },
  Object {
    "detail": "/code-completion/classes",
    "documentation": undefined,
    "filterText": undefined,
    "insertText": undefined,
    "kind": "Class",
    "label": "CodeCompletionClass2",
    "sortText": undefined,
    "textEdit": undefined,
  },
]
`;

exports[`CodeCompletion provideCompletionItems() should provide a completion list for default imports 1`] = `
Array [
  Object {
    "detail": "/code-completion/classes",
    "documentation": undefined,
    "filterText": undefined,
    "insertText": undefined,
    "kind": "File",
    "label": "classes",
    "sortText": undefined,
    "textEdit": undefined,
  },
]
`;

exports[`CodeCompletion provideCompletionItems() should provide a completion list for interfaces 1`] = `
Array [
  Object {
    "detail": "/code-completion/classes",
    "documentation": undefined,
    "filterText": undefined,
    "insertText": undefined,
    "kind": "Interface",
    "label": "CodeCompletionInterface",
    "sortText": undefined,
    "textEdit": undefined,
  },
]
`;
